photon-ml/src/main/scala/com/linkedin/photon/ml/function/DiffFunction.scala

```diff
@@ -15,307 +15,38 @@
-   */
-  private def withL1Regularization[Datum <: DataPoint](
-      func: DiffFunction[Datum],
-      regWeight: Double): DiffFunction[Datum]
-    with L1RegularizationTerm =
-      new DiffFunction[Datum] with L1RegularizationTerm {
-
-    override protected[ml] def calculateAt(
-        datum: Datum,
-        coefficients: Vector[Double],
-        cumGradient: Vector[Double]): Double = {
-
-      func.calculateAt(datum, coefficients, cumGradient)
-    }
-
-    override protected[ml] def calculate(
-        data: RDD[Datum],
-        broadcastedCoefficients: Broadcast[Vector[Double]]): (Double, Vector[Double]) = {
-
-      func.calculate(data, broadcastedCoefficients)
-    }
-
-    override protected[ml] def calculate(
-        data: Iterable[Datum],
-        coefficients: Vector[Double]): (Double, Vector[Double]) = {
-
-      func.calculate(data, coefficients)
-    }
-
-    override def getL1RegularizationParam: Double = regWeight
-  }
-
-  /**
-   * Add regularization to the differentiable function. Under the hood, the L2 regularization part is added to the loss
-   * function values/gradient/Hessian, but the L1 regularization has only a regularization weight to be further used by
-   * the optimizer (especially
-   * [[[http://www.scalanlp.org/api/breeze/index.html#breeze.optimize.OWLQN breeze.optimize.OWLQN]] used in
-   * [[LBFGS LBFGS]]).
-   * @param func The differentiable function
-   * @param regularizationContext The regularization context
-   * @param regWeight The regularization weight
-   * @tparam Datum Datum type
-   * @return The differentiable function with necessary decorations
-   */
-  def withRegularization[Datum <: DataPoint](
-      func: DiffFunction[Datum],
-      regularizationContext: RegularizationContext,
-      regWeight: Double): DiffFunction[Datum] = {
-
-    val (l1Weight, l2Weight) = (
-      regularizationContext.getL1RegularizationWeight(regWeight),
-      regularizationContext.getL2RegularizationWeight(regWeight))
-
-    (l1Weight,  l2Weight) match {
-      case (0.0, 0.0) =>
-        // No regularization
-        func
-      case (0.0, l2) =>
-        // Only l2 regularization
-        withL2Regularization(func, l2)
-      case (l1, 0.0) =>
-        // Only l1 regularization
-        withL1Regularization(func, l1)
-      case (l1, l2) =>
-        // L1 and L2 regularization
-        val funcWithL2 = withL2Regularization(func, l2)
-        withL1Regularization(funcWithL2, l1)
-    }
-  }
+  * Trait for differentiable function. The value/gradient of a differentiable function depends on the model coefficients
+  * (e.g., the coefficients learned in a generalized linear model).
+  *
+  * FAQ: Why do the functions take an ObjectProvider[ Vector[Double] ] for coefficients instead of a Vector[Double]?
+  *      Why may the functions potentially receive a Broadcast[ Vector[Double] ]?
+  * A: In the optimization procedure, the gradient/value/calculate function might be called repeatedly several times
```

Comment:
The FAQ looks confusing to me, please update it.
1. Which functions take ObjectProvider[ Vector[Double] ]?
2. Where is Broadcast[ Vector[Double] ] being used?